### PR TITLE
Create a generic transform type and I/O algorithm

### PR DESCRIPTION
**Pull request by mwoehlke-kitware**
Create a new class to encapsulate a generic (not necessarily linear) transformation. Add an I/O algorithm to load and save these (no implementation at this time). Make `homography` a subclass of the new `transform`.

### CODE REVIEW

**Comment by mwoehlke-kitware** [comment]
Sigh... apparently my attempt to be clever and arrange for `clone` to return the type of the pointer you already have (i.e. if you have a `homography`, you will get back a `homography`, rather than the base `transform` type) works... on GCC ≥ 7 (which is why it built for me, locally), but not on MSVC or clang (any version). (I'm also not sure it's allowed by the standard.)

Would we rather:
- Not make `homography` a subclass of `transform` for now?
- Require users to do the downcast?
- Something else?

File: vital/algo/transform_io.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  /// Implementation specific load functionality.
+  /**
+   * Concrete implementations of transform_io class must provide an
+   * implementation for this method.
+   *
+   * \param filename the path to the file the load
+   * \returns a transform instance referring to the loaded transform
+   */
+  virtual kwiver::vital::transform_sptr load_(
+    std::string const& filename ) const = 0;
```
Comment (linus-sherrill) [inline vital/algo/transform_io.h]: Would load_ and save_ be more useful if they operated on a stream rather than a file name?

Comment (mwoehlke-kitware) [reply vital/algo/transform_io.h]: Maybe, maybe not... that would give the flexibility to operate on something that is not a file, but only if the implementation is able to operate on a `std::istream`.

At any rate, this is copied from `image_io`...

File: vital/types/transform.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+namespace vital {
+
+// Forward declarations of abstract transform class
+class transform;
+// typedef for a transform shared pointer
+typedef std::shared_ptr< transform > transform_sptr;
+
+// ============================================================================
+/// Abstract base transformation representation class
+class VITAL_EXPORT transform
```
Comment (mleotta) [inline vital/types/transform.h]: This is specifically a 2D transform.  We also have 3D transforms in KWIVER.  Should this be called `transform_2` and should we add a `transform_3` for `vector_3d`?  Possibly it should be `transform_2d`, but note that the `d` in `vector_2d` stands for `double`.  There is also `vector_2f`, `vector_2i`, etc. but the higher level classes almost exclusively use the `double` variants.

Comment (mwoehlke-kitware) [reply vital/types/transform.h]: Possibly, though as you seem to notice, what to name it isn't obvious. I was inclined to punt on that until it matters, and rename later if needed.

Comment (mwoehlke-kitware) [reply vital/types/transform.h]: Okay, per F2F discussion, the class is now `transform_2d`. Note that the `d` does actually refer to `double`, in that `map` operates on vectors whose scalar type is `double`.

**Pull request merged**